feat: Update data collection rule to support `AgentDirectToStore` kind

TrentSteenholdt-Insight opened the pull request:
## Description

~- Updated data collection rule properties to use API version '2024-03-11'.~
- Added support for 'AgentDirectToStore' data collection rule type. 
- Adjusted resource definitions to accommodate the new properties and types.

### Updates

- @krbar kind of took his PR and did half the work in #6281. However as he requested, I've added the E2E for this and bumped the version again to 0.10.0 following his PR

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
| [![avm.res.insights.data-collection-rule](https://github.com/TrentSteenholdt-Insight/bicep-registry-modules/actions/workflows/avm.res.insights.data-collection-rule.yml/badge.svg)](https://github.com/TrentSteenholdt-Insight/bicep-registry-modules/actions/workflows/avm.res.insights.data-collection-rule.yml) | 


## Type of Change

- Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation
- [ ] Update to CI Environment or utilities (Non-module affecting changes)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] I have updated the module's CHANGELOG.md file with an entry for the next version

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

TrentSteenholdt-Insight committed:
AVM-RES: Update data collection rule to support new API version

- Updated data collection rule properties to use API version '2024-03-11'.
- Added support for 'AgentDirectToStore' data collection rule type.
- Introduced new types for 'AgentDirectToStore' and its destinations.
- Adjusted resource definitions to accommodate the new properties and types.

TrentSteenholdt-Insight committed:
AVM-RES: Update API version for data collection rules

- Updated API version for `Microsoft.Insights/dataCollectionRules` from 2023-03-11 to 2024-03-11 in all relevant files.
- Incremented version number in `version.json` to 0.9.
- Adjusted resource references in `main.bicep`, `main.json`, and `nested_lock.bicep` to reflect the new API version.

TrentSteenholdt-Insight committed:
fix: Update changelog for data collection rule

- Added support for preview feature `AgentDirectToStore` kind type in version 0.9.0.

TrentSteenholdt-Insight committed:
feat: Updated AVM module with missing prereqs

TrentSteenholdt-Insight committed:
AVM-RES: Add support for kind `AgentDirectToStore`

- Updated CHANGELOG.md to include new feature.
- Bumped version to 0.9 in version.json.

TrentSteenholdt-Insight committed:
Merge branch 'main' of https://github.com/Azure/bicep-registry-modules

TrentSteenholdt-Insight committed:
Merge branch 'main' of https://github.com/TrentSteenholdt-Insight/bicep-registry-modules into feat/add-AgentDirectToStore

TrentSteenholdt-Insight committed:
AVM-RES: Add support for 'AgentDirectToStore' kind

- Updated CHANGELOG.md to include 'AgentDirectToStore' support.
- Enhanced README.md with examples and details for 'AgentDirectToStore'.
- Modified main.json to define 'AgentDirectToStore' properties.
- Updated version.json to reflect version 0.10.

TrentSteenholdt-Insight committed:
AVM-RES: Add AgentDirectToStore E2E test modules

- Introduced new Bicep files for E2E testing of AgentDirectToStore functionality.
- Created 'dependencies.bicep' to define required resources for the test.
- Developed 'main.test.bicep' to deploy and validate the AgentDirectToStore setup.
- Ensured proper parameterization for resource deployment and outputs.

TrentSteenholdt-Insight committed:
AVM-RES: Update parameters for nested dependencies module

- Refactored parameters in the nestedDependencies module to use storageAccountName and eventHubNamespaceName.
- Ensured consistency in naming conventions for resource deployment.

TrentSteenholdt-Insight committed:
AVM-RES: Update E2E test for AgentDirectToStore

- Enhanced the E2E test module for AgentDirectToStore.
- Improved deployment parameters for better resource management.
- Ensured compatibility with the latest module updates.

TrentSteenholdt-Insight committed:
AVM-RES: Update README for AgentDirectToStore examples

- Added new usage examples for sending data to Azure Blob Storage using the AgentDirectToStore kind.
- Updated example numbering for consistency across the README.
- Enhanced details for the Bicep and JSON parameter files for better clarity.

TrentSteenholdt-Insight committed:
AVM-RES: Update README and test metadata for AgentDirectToStore

- Updated README examples to reflect changes in data sending methods.
- Changed test metadata to align with the new example description.
- Adjusted references to Azure Storage or Event Hub in documentation.

TrentSteenholdt-Insight committed:
AVM-RES: Update data collection rule properties for AgentDirectToStore

- Added 'AgentDirectToStore' to the list of supported kinds in dataCollectionRulePropertiesUnion.
- Updated the logic to handle data flows and destinations for 'AgentDirectToStore'.

TrentSteenholdt-Insight committed:
AVM-RES: Update README and test metadata for AgentDirectToStore

- Updated README to reflect changes in the AgentDirectToStore module.
- Modified test metadata for improved clarity and accuracy.

TrentSteenholdt-Insight commented:
@krbar bump. I would like this ASAP so I can use it in my code.

TrentSteenholdt-Insight committed:
Merge branch 'main' into feat/add-AgentDirectToStore

TrentSteenholdt-Insight commented:
Bump. @AlexanderSehr could you possibly help. You helped last time when little PRs just got ghosted.

AlexanderSehr commented on avm/res/insights/data-collection-rule/tests/e2e/agent-direct/main.test.bicep:
```diff
@@ -0,0 +1,128 @@
+    name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+    params: {
+      name: '${namePrefix}${serviceShort}001'
+      location: resourceLocation
```
```suggestion
```

AlexanderSehr reviewed (commented):
Hey @TrentSteenholdt-Insight,
great work as always. Just a few small comments. Once addressed I'm happy to merge and would be suprised if @krbar objects given the change.
Please remember to refresh the template files after applying the changes/comments (if you agree with them)

TrentSteenholdt-Insight committed:
Update avm/res/insights/data-collection-rule/tests/e2e/agent-direct/main.test.bicep

Co-authored-by: Alexander Sehr <ASehr@hotmail.de>

TrentSteenholdt-Insight committed:
Update avm/res/insights/data-collection-rule/tests/e2e/agent-direct/main.test.bicep

Co-authored-by: Alexander Sehr <ASehr@hotmail.de>

TrentSteenholdt-Insight committed:
Update avm/res/insights/data-collection-rule/tests/e2e/agent-direct/main.test.bicep

Co-authored-by: Alexander Sehr <ASehr@hotmail.de>

TrentSteenholdt-Insight committed:
Update avm/res/insights/data-collection-rule/tests/e2e/agent-direct/main.test.bicep

Co-authored-by: Alexander Sehr <ASehr@hotmail.de>

TrentSteenholdt-Insight committed:
CHANGE: Update description for agentDirectToStoreType

- Modified the description field in agentDirectToStoreType to provide a clearer explanation of its purpose.
- This change enhances the documentation for better understanding of the data collection rule's functionality.

TrentSteenholdt-Insight committed:
Merge branch 'main' into feat/add-AgentDirectToStore
